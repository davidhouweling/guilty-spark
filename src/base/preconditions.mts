export const Preconditions = {
  checkExists<T>(value: T, errorMessage?: string): NonNullable<T> {
    if (value == null) {
      throw new Error(errorMessage ?? "Value cannot be null");
    }

    return value;
  },
};
