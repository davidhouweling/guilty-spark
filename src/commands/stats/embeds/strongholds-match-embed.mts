import { GameVariantCategory } from "halo-infinite-api";
import { BaseMatchEmbed, PlayerStats } from "./base-match-embed.mjs";

export class StrongholdsMatchEmbed extends BaseMatchEmbed<GameVariantCategory.MultiplayerStrongholds> {
  override getPlayerObjectiveStats(
    stats: PlayerStats<GameVariantCategory.MultiplayerStrongholds>,
  ): Map<string, string> {
    return new Map([
      ["Captures", stats.ZonesStats.StrongholdCaptures.toString()],
      ["Occupation time", this.haloService.getReadableDuration(stats.ZonesStats.StrongholdOccupationTime)],
      ["Secures", stats.ZonesStats.StrongholdSecures.toString()],
      ["Offensive kills", stats.ZonesStats.StrongholdOffensiveKills.toString()],
      ["Defensive kills", stats.ZonesStats.StrongholdDefensiveKills.toString()],
    ]);
  }
}
