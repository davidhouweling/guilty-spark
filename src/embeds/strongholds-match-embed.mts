import type { GameVariantCategory, Stats } from "halo-infinite-api";
import type { EmbedPlayerStats } from "./base-match-embed.mjs";
import { BaseMatchEmbed, StatsValueSortBy } from "./base-match-embed.mjs";

export class StrongholdsMatchEmbed extends BaseMatchEmbed<GameVariantCategory.MultiplayerStrongholds> {
  override getPlayerObjectiveStats(stats: Stats<GameVariantCategory.MultiplayerStrongholds>): EmbedPlayerStats {
    return new Map([
      ["Captures", { value: stats.ZonesStats.StrongholdCaptures, sortBy: StatsValueSortBy.DESC }],
      [
        "Occupation time",
        {
          value: this.haloService.getDurationInSeconds(stats.ZonesStats.StrongholdOccupationTime),
          sortBy: StatsValueSortBy.DESC,
          display: this.haloService.getReadableDuration(stats.ZonesStats.StrongholdOccupationTime, this.locale),
        },
      ],
      ["Secures", { value: stats.ZonesStats.StrongholdSecures, sortBy: StatsValueSortBy.DESC }],
      ["Offensive kills", { value: stats.ZonesStats.StrongholdOffensiveKills, sortBy: StatsValueSortBy.DESC }],
      ["Defensive kills", { value: stats.ZonesStats.StrongholdDefensiveKills, sortBy: StatsValueSortBy.DESC }],
    ]);
  }
}
