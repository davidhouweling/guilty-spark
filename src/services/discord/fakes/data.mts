import type {
  APIApplicationCommandInteraction,
  APIMessage,
  APIMessageComponentButtonInteraction,
  APIModalSubmitInteraction,
  APIPingInteraction,
  RESTPostAPIChannelMessagesThreadsResult,
} from "discord-api-types/v10";
import {
  ButtonStyle,
  ComponentType,
  MessageFlags,
  ApplicationCommandOptionType,
  ChannelType,
  GuildMemberFlags,
  InteractionType,
  Locale,
  MessageType,
} from "discord-api-types/v10";
import type { QueueData } from "../discord.mjs";

export const apiMessage: APIMessage = {
  type: MessageType.Default,
  content: "Hello, world!",
  mentions: [],
  mention_roles: [],
  attachments: [],
  embeds: [],
  timestamp: "2024-12-06T12:03:09.182000+00:00",
  edited_timestamp: null,
  components: [],
  id: "1314562775950954626",
  channel_id: "1299532381308325949",
  author: {
    id: "000000000000000001",
    username: "soundmanD",
    avatar: "e803b2f163fda5aeba2cf4820e3a6535",
    discriminator: "0850",
    global_name: null,
  },
  mention_everyone: false,
  pinned: false,
  tts: false,
};

export const pingInteraction: APIPingInteraction = {
  id: "fake-id",
  type: InteractionType.Ping,
  application_id: "fake-application-id",
  token: "fake-token",
  version: 1,
  app_permissions: "",
  authorizing_integration_owners: {},
  entitlements: [],
};

const fakeBaseInteraction = {
  app_permissions: "fake-permissions",
  application_id: "fake-application-id",
  authorizing_integration_owners: {},
  context: 0,
  entitlements: [],
  id: "fake-id",
  locale: Locale.EnglishUS,
  member: {
    deaf: false,
    joined_at: "2024-05-11T11:45:17.722000+00:00",
    mute: false,
    permissions: "fake-permissions",
    roles: [],
    user: {
      avatar: null,
      discriminator: "0",
      global_name: null,
      id: "fake-user-id",
      username: "soundmanD",
    },
    flags: GuildMemberFlags.CompletedOnboarding,
  },
  token: "fake-token",
};

const fakeBaseAPIApplicationCommandInteraction: Omit<APIApplicationCommandInteraction, "type" | "data"> = {
  ...fakeBaseInteraction,
  channel: {
    guild_id: "fake-channel-guild-id",
    id: "fake-channel-id",
    type: ChannelType.GuildText,
  },
  channel_id: "fake-channel-id",
  version: 1,
};

export const applicationCommandInteractionStatsNeatQueue: APIApplicationCommandInteraction = {
  ...fakeBaseAPIApplicationCommandInteraction,
  type: InteractionType.ApplicationCommand,
  data: {
    id: "1296081783443685377",
    name: "stats",
    options: [
      {
        name: "neatqueue",
        options: [
          {
            name: "channel",
            type: 7,
            value: "1299532381308325949",
          },
          {
            name: "queue",
            type: 4,
            value: 1418,
          },
        ],
        type: 1,
      },
    ],
    resolved: {
      channels: {
        "1299532381308325949": {
          id: "1299532381308325949",
          name: "ü•âresults",
          permissions: "2230813650837056",
          type: 0,
        },
      },
    },
    type: 1,
  },
};

export const applicationCommandInteractionStatsMatch: APIApplicationCommandInteraction = {
  ...fakeBaseAPIApplicationCommandInteraction,
  type: InteractionType.ApplicationCommand,
  data: {
    id: "1300004385459408960",
    name: "stats",
    options: [
      {
        name: "match",
        options: [
          {
            name: "id",
            type: ApplicationCommandOptionType.String,
            value: "d81554d7-ddfe-44da-a6cb-000000000ctf",
          },
        ],
        type: 1,
      },
    ],
    type: 1,
  },
};

export const buttonClickInteraction: APIMessageComponentButtonInteraction = {
  ...fakeBaseInteraction,
  channel: {
    guild_id: "fake-channel-guild-id",
    id: "fake-channel-id",
    last_message_id: "fake-last-message-id",
    name: "general",
    nsfw: false,
    parent_id: "fake-parent-id",
    position: 0,
    rate_limit_per_user: 0,
    topic: null,
    type: ChannelType.GuildText,
  },
  channel_id: "fake-channel-id",
  data: { component_type: ComponentType.Button, custom_id: "btn_yes" },
  message: {
    application_id: "fake-application-id",
    attachments: [],
    author: {
      avatar: null,
      avatar_decoration_data: null,
      bot: true,
      discriminator: "2015",
      global_name: null,
      id: "fake-guilty-spark-id",
      public_flags: 524288,
      username: "Guilty Spark",
    },
    channel_id: "fake-channel-id",
    components: [
      {
        components: [
          {
            custom_id: "btn_yes",
            label: "Yes",
            style: ButtonStyle.Primary,
            type: ComponentType.Button,
          },
          {
            custom_id: "btn_no",
            label: "No",
            style: ButtonStyle.Danger,
            type: ComponentType.Button,
          },
        ],
        type: ComponentType.ActionRow,
      },
    ],
    content: "",
    edited_timestamp: null,
    embeds: [
      {
        description: "Some description",
        fields: [
          {
            inline: false,
            name: "Field title",
            value: "Field value",
          },
        ],
        title: "Embed title",
      },
    ],
    flags: MessageFlags.Ephemeral,
    id: "fake-message-id",
    interaction: {
      id: "fake-interaction-id",
      name: "fake-interaction-name",
      type: InteractionType.ApplicationCommand,
      user: {
        avatar: null,
        avatar_decoration_data: null,
        discriminator: "0",
        global_name: "fake-user-global-name",
        id: "fake-user-id",
        username: "fake-username",
      },
    },
    interaction_metadata: {
      authorizing_integration_owners: {},
      id: "fake-interaction-metadata-id",
      type: InteractionType.ApplicationCommand,
      user: {
        avatar: null,
        avatar_decoration_data: null,
        discriminator: "0",
        global_name: "fake-user-global-name",
        id: "fake-user-id",
        username: "fake-username",
      },
    },
    mention_everyone: false,
    mention_roles: [],
    mentions: [],
    pinned: false,
    position: 0,
    timestamp: "2025-01-10T09:42:25.495000+00:00",
    tts: false,
    type: 20,
    webhook_id: "fake-webhook-id",
  },
  type: InteractionType.MessageComponent,
  version: 1,
};

export const modalSubmitInteraction: APIModalSubmitInteraction = {
  ...fakeBaseInteraction,
  channel: {
    guild_id: "fake-channel-guild-id",
    id: "fake-channel-id",
    last_message_id: "fake-last-message-id",
    name: "general",
    nsfw: false,
    parent_id: "fake-parent-id",
    position: 0,
    rate_limit_per_user: 0,
    topic: null,
    type: ChannelType.GuildText,
  },
  channel_id: "fake-channel-id",
  data: {
    components: [
      {
        components: [{ custom_id: "text_input", type: ComponentType.TextInput, value: "Hello!" }],
        type: 1,
      },
    ],
    custom_id: "text_input_modal",
  },
  message: {
    application_id: "fake-application-id",
    attachments: [],
    author: {
      avatar: null,
      avatar_decoration_data: null,
      bot: true,
      discriminator: "2015",
      global_name: null,
      id: "fake-guilty-spark-id",
      public_flags: 524288,
      username: "Guilty Spark",
    },
    channel_id: "fake-channel-id",
    components: [
      {
        components: [
          {
            custom_id: "btn_modal",
            label: "Modal",
            style: ButtonStyle.Secondary,
            type: ComponentType.Button,
          },
        ],
        type: ComponentType.ActionRow,
      },
    ],
    content: "",
    edited_timestamp: null,
    embeds: [],
    flags: 64,
    id: "fake-message-id",
    interaction: {
      id: "fake-interaction-id",
      name: "fake-interaction-name",
      type: InteractionType.ApplicationCommand,
      user: {
        avatar: null,
        avatar_decoration_data: null,
        discriminator: "0",
        global_name: "fake-user-global-name",
        id: "fake-user-id",
        username: "fake-username",
      },
    },
    interaction_metadata: {
      authorizing_integration_owners: {},
      id: "fake-interaction-metadata-id",
      type: InteractionType.ApplicationCommand,
      user: {
        avatar: null,
        avatar_decoration_data: null,
        discriminator: "0",
        global_name: "fake-user-global-name",
        id: "fake-user-id",
        username: "fake-username",
      },
    },
    mention_everyone: false,
    mention_roles: [],
    mentions: [],
    pinned: false,
    position: 0,
    timestamp: "2025-01-24T09:41:38.252000+00:00",
    tts: false,
    type: 20,
    webhook_id: "fake-webhook-id",
  },
  type: InteractionType.ModalSubmit,
  version: 1,
};

function aFakeNeatQueueMessageWith(opts: Partial<APIMessage> = {}): APIMessage {
  return {
    type: MessageType.Default,
    content: "",
    mentions: [],
    mention_roles: [],
    attachments: [],
    embeds: opts.embeds ?? [
      {
        title: "üèÜ Winner For Queue#1 üèÜ",
        color: 16711680,
        timestamp: "2024-12-06T12:03:08.786000+00:00",
        fields: [
          {
            name: "Eagle",
            value:
              "<@000000000000000001> *-30.3* **(993.2)**\n<@000000000000000002> *-30.1* **(978.5)**\n<@000000000000000003> *-30.0* **(966.1)**\n<@000000000000000004> *-29.6* **(926.6)**",
            inline: true,
          },
          {
            name: "__Cobra__",
            value:
              "<@000000000000000005> *+28.1* **(1221.0)**\n<@000000000000000006> *+29.5* **(1081.8)**\n<@000000000000000007> *+29.7* **(1063.2)**\n<@000000000000000008> *+32.8* **(755.4)**",
            inline: true,
          },
        ],
      },
    ],
    timestamp: "2024-12-06T12:03:09.182000+00:00",
    edited_timestamp: null,
    components: [],
    id: "1314562775950954626",
    channel_id: "1299532381308325949",
    author: {
      id: "857633321064595466",
      username: "NeatQueue",
      avatar: "e803b2f163fda5aeba2cf4820e3a6535",
      discriminator: "0850",
      public_flags: 65536,
      flags: 65536,
      bot: true,
      banner: null,
      accent_color: null,
      global_name: null,
      avatar_decoration_data: null,
    },
    pinned: false,
    mention_everyone: false,
    tts: false,
    ...opts,
  };
}

export const channelMessages: APIMessage[] = [
  aFakeNeatQueueMessageWith({
    embeds: [
      {
        title: "üèÜ Winner For Queue#10 üèÜ",
        color: 16711680,
        timestamp: "2024-12-05T11:52:20.193000+00:00",
        fields: [
          {
            name: "__Eagle__",
            value:
              "<@000000000000000001> *+29.8* **(1235.8)**\n<@000000000000000002> *+31.6* **(1052.1)**\n<@000000000000000003> *+32.5* **(983.2)**\n<@000000000000000004> *+32.6* **(956.2)**",
            inline: true,
          },
          {
            name: "Cobra",
            value:
              "<@000000000000000005> *-32.5* **(1117.3)**\n<@000000000000000006> *-32.2* **(1099.1)**\n<@000000000000000007> *-31.5* **(1036.6)**\n<@000000000000000008> *-31.3* **(1012.9)**",
            inline: true,
          },
        ],
      },
    ],
    timestamp: "2024-12-05T11:52:21.235000+00:00",
    id: "1314197670398525493",
  }),
  aFakeNeatQueueMessageWith({
    embeds: [
      {
        title: "üèÜ Winner For Queue#7 üèÜ",
        color: 16711680,
        timestamp: "2024-12-06T11:05:39.576000+00:00",
        fields: [
          {
            name: "Eagle",
            value:
              "<@000000000000000009> *-28.2* **(1033.4)**\n<@000000000000000008> *-28.1* **(1023.5)**\n<@000000000000000007> *-27.9* **(1008.7)**\n<@000000000000000006> *-27.8* **(996.1)**",
            inline: true,
          },
          {
            name: "__Cobra__",
            value:
              "<@000000000000000010> *+26.4* **(1289.1)**\n<@000000000000000011> *+27.0* **(1262.8)**\n<@000000000000000012> *+29.1* **(1054.5)**\n<@000000000000000013> *+29.5* **(1052.3)**",
            inline: true,
          },
        ],
      },
    ],
    timestamp: "2024-12-06T11:05:41.120000+00:00",
    id: "1314548313730846792",
  }),
  aFakeNeatQueueMessageWith({
    embeds: [
      {
        title: "üèÜ Winner For Queue#4 üèÜ",
        color: 16711680,
        timestamp: "2024-12-06T09:55:16.814000+00:00",
        fields: [
          {
            name: "Eagle",
            value:
              "<@000000000000000010> *-29.6* **(1057.0)**\n<@000000000000000008> *-29.3* **(1033.2)**\n<@000000000000000005> *-29.2* **(1022.9)**\n<@000000000000000001> *-28.9* **(992.6)**",
            inline: true,
          },
          {
            name: "__Cobra__",
            value:
              "<@000000000000000012> *+27.8* **(1262.7)**\n<@000000000000000014> *+28.4* **(1203.9)**\n<@000000000000000016> *+29.2* **(1128.3)**\n<@000000000000000018> *+31.7* **(880.2)**",
            inline: true,
          },
        ],
      },
    ],
    timestamp: "2024-12-06T09:55:17.673000+00:00",
    id: "1314530599309934632",
  }),
  aFakeNeatQueueMessageWith({
    embeds: [
      {
        title: "üèÜ Winner For Queue#3 üèÜ",
        color: 16711680,
        timestamp: "2024-12-06T08:24:07.415000+00:00",
        fields: [
          {
            name: "__Eagle__",
            value:
              "<@000000000000000001> *+29.2* **(1234.9)**\n<@000000000000000003> *+29.8* **(1175.5)**\n<@000000000000000005> *+31.0* **(1062.6)**\n<@000000000000000007> *+32.1* **(951.3)**",
            inline: true,
          },
          {
            name: "Cobra",
            value:
              "<@000000000000000002> *-32.2* **(1231.7)**\n<@000000000000000004> *-31.0* **(1114.8)**\n<@000000000000000006> *-30.7* **(1086.6)**\n<@000000000000000008> *-28.3* **(848.5)**",
            inline: true,
          },
        ],
      },
    ],
    timestamp: "2024-12-06T08:24:08.804000+00:00",
    id: "1314507661211074611",
  }),
  aFakeNeatQueueMessageWith(),
  aFakeNeatQueueMessageWith({
    embeds: [
      {
        color: 16711680,
        timestamp: "2024-12-06T06:24:07.415000+00:00",
        fields: [
          {
            name: "Some fake value",
            value: "just doing this to for tests",
            inline: true,
          },
        ],
      },
    ],
  }),
  {
    type: MessageType.Default,
    content: "",
    mentions: [],
    mention_roles: [],
    attachments: [],
    embeds: [],
    timestamp: "2024-12-06T12:03:09.182000+00:00",
    edited_timestamp: null,
    components: [],
    id: "1314562775950954626",
    channel_id: "1299532381308325949",
    author: {
      id: "000000000000000001",
      username: "soundmanD",
      avatar: "e803b2f163fda5aeba2cf4820e3a6535",
      discriminator: "0850",
      public_flags: 65536,
      flags: 65536,
      global_name: null,
    },
    pinned: false,
    mention_everyone: false,
    tts: false,
  },
];

export const discordNeatQueueData: QueueData = {
  message: {
    type: 0,
    content: "",
    mentions: [],
    mention_roles: [],
    attachments: [],
    embeds: [
      {
        title: "üèÜ Winner For Queue#777 üèÜ",
        color: 16711680,
        timestamp: "2024-11-26T11:30:00.000000+00:00",
        fields: [
          {
            name: "Eagle",
            value:
              "<@000000000000000001> *-31.5* **(1131.7)**\n" +
              "<@000000000000000002> *-31.4* **(1117.9)**\n" +
              "<@000000000000000003> *-30.4* **(1017.7)**\n" +
              "<@000000000000000004> *-29.8* **(966.3)**",
            inline: true,
          },
          {
            name: "__Cobra__",
            value:
              "<@000000000000000005> *+29.6* **(1195.0)**\n" +
              "<@000000000000000006> *+29.8* **(1185.0)**\n" +
              "<@000000000000000007> *+30.9* **(1067.3)**\n" +
              "<@000000000000000008> *+32.8* **(887.7)**",
            inline: true,
          },
        ],
      },
    ],
    timestamp: "2024-11-26T11:30:00.000000+00:00",
    edited_timestamp: null,
    components: [],
    id: "1310523001611096064",
    channel_id: "1299532381308325949",
    author: {
      id: "857633321064595466",
      username: "NeatQueue",
      avatar: "e803b2f163fda5aeba2cf4820e3a6535",
      discriminator: "0850",
      public_flags: 65536,
      flags: 65536,
      bot: true,
      global_name: null,
    },
    pinned: false,
    mention_everyone: false,
    tts: false,
  },
  timestamp: new Date("2024-11-26T11:30:00.000Z"),
  teams: [
    {
      name: "Eagle",
      players: [
        {
          id: "000000000000000001",
          username: "discord_user_01",
          avatar: "157e517cdbf371a47aaead44675714a3",
          discriminator: "0",
          global_name: "DiscordUser01",
        },
        {
          id: "000000000000000002",
          username: "discord_user_02",
          avatar: null,
          discriminator: "0",
          global_name: "DiscordUser02",
        },
        {
          id: "000000000000000003",
          username: "discord_user_03",
          avatar: null,
          discriminator: "0",
          global_name: null,
        },
        {
          id: "000000000000000004",
          username: "discord_user_04",
          avatar: "6f9ef56a174047263d9c81e9b2559fdc",
          discriminator: "0",
          global_name: "DiscordUser04",
        },
      ],
    },
    {
      name: "Cobra",
      players: [
        {
          id: "000000000000000005",
          username: "discord_user_05",
          avatar: "4081ed914ba463a1374b2b1b11f7bf60",
          discriminator: "0",
          global_name: "DiscordUser05",
        },
        {
          id: "000000000000000006",
          username: "discord_user_06",
          avatar: null,
          discriminator: "0",
          global_name: "DiscordUser06",
        },
        {
          id: "000000000000000007",
          username: "discord_user_07",
          avatar: "d354460036abf8fcddd1282b353bfcd9",
          discriminator: "0",
          global_name: "DiscordUser07",
        },
        {
          id: "000000000000000008",
          username: "discord_user_08",
          avatar: "15c79840f18f08defc64793b55afc2b9",
          discriminator: "0",
          global_name: "DiscordUser08",
        },
      ],
    },
  ],
};

export const channelThreadsResult: RESTPostAPIChannelMessagesThreadsResult = {
  id: "fake-thread-id",
  name: "fake-thread-name",
  type: ChannelType.PublicThread,
  applied_tags: [],
  position: 0,
};
