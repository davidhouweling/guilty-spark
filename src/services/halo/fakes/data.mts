import { MapAsset, MatchStats, PlayerMatchHistory } from "halo-infinite-api";
import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

async function readMatchStats(filename: string): Promise<[string, MatchStats]> {
  try {
    const fileContents = await readFile(path.join(__dirname, "data", filename), "utf-8");
    const matchStats = JSON.parse(fileContents) as MatchStats;
    return [matchStats.MatchId, matchStats];
  } catch (error) {
    console.error(`Failed to read match stats from ${filename}: ${error as Error}`);
    throw error;
  }
}

async function readPlayerMatches(): Promise<PlayerMatchHistory[]> {
  try {
    const fileContents = await readFile(path.join(__dirname, "data", "player-matches.json"), "utf-8");
    return JSON.parse(fileContents) as PlayerMatchHistory[];
  } catch (error) {
    console.error(`Failed to read player matches: ${error as Error}`);
    throw error;
  }
}

async function readAssetVersion(): Promise<MapAsset> {
  try {
    const fileContents = await readFile(path.join(__dirname, "data", "asset-version.json"), "utf-8");
    return JSON.parse(fileContents) as MapAsset;
  } catch (error) {
    console.error(`Failed to read asset version: ${error as Error}`);
    throw error;
  }
}

export const matchStats = new Map<string, MatchStats>(
  await Promise.all([
    readMatchStats("ctf.json"),
    readMatchStats("koth.json"),
    readMatchStats("land-grab.json"),
    readMatchStats("oddball.json"),
    readMatchStats("slayer.json"),
    readMatchStats("strongholds.json"),
    readMatchStats("total-control.json"),
  ]),
);

export const playerMatches = await readPlayerMatches();

export const assetVersion = await readAssetVersion();

export const playerXuidsToGametags = new Map([
  ["0100000000000000", "gamertag01"],
  ["0200000000000000", "gamertag02"],
  ["0300000000000000", "gamertag03"],
  ["0400000000000000", "gamertag04"],
  ["0500000000000000", "gamertag05"],
  ["0600000000000000", "gamertag06"],
  ["0700000000000000", "gamertag07"],
  ["0800000000000000", "gamertag08"],
  ["0900000000000000", "gamertag09"],
  ["1000000000000000", "gamertag10"],
  ["1100000000000000", "gamertag11"],
  ["1200000000000000", "gamertag12"],
  ["1300000000000000", "gamertag13"],
  ["1400000000000000", "gamertag14"],
  ["1500000000000000", "gamertag15"],
  ["1600000000000000", "gamertag16"],
]);
