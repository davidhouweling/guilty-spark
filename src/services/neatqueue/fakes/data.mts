import { readFile } from "node:fs/promises";
import path from "node:path";
import { fileURLToPath } from "node:url";
import type {
  NeatQueueJoinQueueRequest,
  NeatQueueLeaveQueueRequest,
  NeatQueueMatchCompletedRequest,
  NeatQueueMatchStartedRequest,
  NeatQueueSubstitutionRequest,
  NeatQueueTeamsCreatedRequest,
} from "../types.mjs";

const __dirname = path.dirname(fileURLToPath(import.meta.url));

async function readFakeData<T>(filename: string): Promise<T> {
  const filePath = path.join(__dirname, "./data", filename);
  const fileContents = await readFile(filePath, "utf-8");
  return JSON.parse(fileContents) as T;
}

export const joinQueueData = await readFakeData<NeatQueueJoinQueueRequest>("join-queue.json");
export const leaveQueueData = await readFakeData<NeatQueueLeaveQueueRequest>("leave-queue.json");
export const matchStartedData = await readFakeData<NeatQueueMatchStartedRequest>("match-started.json");
export const teamsCreatedData = await readFakeData<NeatQueueTeamsCreatedRequest>("teams-created.json");
export const substitutionData = await readFakeData<NeatQueueSubstitutionRequest>("substitution.json");
export const matchCompletedData = await readFakeData<NeatQueueMatchCompletedRequest>("match-completed.json");
