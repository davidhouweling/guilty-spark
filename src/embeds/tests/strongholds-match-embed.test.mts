import { describe, beforeEach, expect, it } from "vitest";
import { matchStats, playerXuidsToGametags } from "../../services/halo/fakes/data.mjs";
import { StrongholdsMatchEmbed } from "../strongholds-match-embed.mjs";
import type { HaloService } from "../../services/halo/halo.mjs";
import { Preconditions } from "../../base/preconditions.mjs";
import { aFakeHaloServiceWith } from "../../services/halo/fakes/halo.fake.mjs";
import type { DiscordService } from "../../services/discord/discord.mjs";
import { aFakeDiscordServiceWith } from "../../services/discord/fakes/discord.fake.mjs";
import type { GuildConfigRow } from "../../services/database/types/guild_config.mjs";
import { aFakeGuildConfigRow } from "../../services/database/fakes/database.fake.mjs";

const match = Preconditions.checkExists(matchStats.get("099deb74-3f60-48cf-8784-0strongholds"));

describe("StrongholdsMatchEmbed", () => {
  const locale = "en-US";
  let discordService: DiscordService;
  let haloService: HaloService;
  let guildConfig: GuildConfigRow;
  let matchEmbed: StrongholdsMatchEmbed;

  beforeEach(() => {
    discordService = aFakeDiscordServiceWith();
    haloService = aFakeHaloServiceWith();
    guildConfig = aFakeGuildConfigRow();
    matchEmbed = new StrongholdsMatchEmbed({ discordService, haloService, guildConfig, locale });
  });

  describe("getEmbed", () => {
    it("returns the expected embed", async () => {
      const result = await matchEmbed.getEmbed(match, playerXuidsToGametags);
      expect(result).toMatchSnapshot();
    });
  });
});
