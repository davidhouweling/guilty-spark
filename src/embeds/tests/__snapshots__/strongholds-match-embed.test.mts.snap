// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`StrongholdsMatchEmbed > getEmbed > returns the expected embed 1`] = `
{
  "description": "-# Legend: **Best in team** | __**Best overall**__",
  "fields": [
    {
      "inline": false,
      "name": "Eagle",
      "value": "Team Score: 175 | Team K:D:A: 62:83:24",
    },
    {
      "inline": true,
      "name": "gamertag01 (Rank: 5 | Score: 2,780)",
      "value": "Kills: **20**
Deaths: 21
Assists: 4
KDA: **0.33**
Headshot kills: 11
Shots H:F (acc): 125:201 __**(62.18%)**__
Damage D:T (D/T): 5,771:**5,949** **(0.97)**
Average life duration (damage/life): 22s (274.81)
Captures: 10
Occupation time: 1m 34s
Secures: 2
Offensive kills: 0
Defensive kills: **3**
2x<:ShotCaller:1322884687693086771> 3x<:Perfect:1322879561515532308> 2x<:DoubleKill:1322814147980689430> 2x<:Reversal:1322872797701083166> <:Rifleman:1322860656399220828>",
    },
    {
      "inline": true,
      "name": "gamertag02 (Rank: 5 | Score: 2,700)",
      "value": "Kills: 17
Deaths: 22
Assists: **10**
KDA: -1.66
Headshot kills: **12**
Shots H:F (acc): __**155**__:__**253**__ (61.26%)
Damage D:T (D/T): **6,633**:6,892 (0.96)
Average life duration (damage/life): 21s **(301.5)**
Captures: 9
Occupation time: 1m 28s
Secures: 2
Offensive kills: **3**
Defensive kills: **3**
<:Perfect:1322879561515532308> 3x<:DoubleKill:1322814147980689430> 2x<:Reversal:1322872797701083166> <:LastShot:1322884170606972959> <:Rifleman:1322860656399220828> <:Stick:1322871787695898676> <:Wingman:1322854221095108609>",
    },
    {
      "inline": false,
      "name": "
",
      "value": "
",
    },
    {
      "inline": true,
      "name": "gamertag06 (Rank: 5 | Score: 2,230)",
      "value": "Kills: 13
Deaths: 21
Assists: 5
KDA: -6.33
Headshot kills: 9
Shots H:F (acc): 88:170 (51.76%)
Damage D:T (D/T): 4,528:6,405 (0.71)
Average life duration (damage/life): 21s (215.62)
Captures: __**12**__
Occupation time: __**2m 10s**__
Secures: 2
Offensive kills: 1
Defensive kills: **3**
6x<:ShotCaller:1322884687693086771> <:Perfect:1322879561515532308> <:Snipe:1322872969571340370> 2x<:Killjoy:1322802841039147129> <:Rifleman:1322860656399220828>",
    },
    {
      "inline": true,
      "name": "gamertag08 (Rank: 5 | Score: 1,925)",
      "value": "Kills: 12
Deaths: **19**
Assists: 7
KDA: -4.66
Headshot kills: 6
Shots H:F (acc): 128:220 (58.18%)
Damage D:T (D/T): 5,716:6,138 (0.93)
Average life duration (damage/life): **24s** (300.84)
Captures: 6
Occupation time: 1m 10s
Secures: __**3**__
Offensive kills: 1
Defensive kills: 2
<:RemoteDetonation:1322883672470458380> <:Perfect:1322879561515532308> <:Rifleman:1322860656399220828>",
    },
    {
      "inline": false,
      "name": "
",
      "value": "
",
    },
    {
      "inline": false,
      "name": "Cobra",
      "value": "Team Score: 250 | Team K:D:A: 83:62:32",
    },
    {
      "inline": true,
      "name": "gamertag07 (Rank: 1 | Score: 3,550)",
      "value": "Kills: __**30**__
Deaths: 20
Assists: 3
KDA: __**11**__
Headshot kills: __**20**__
Shots H:F (acc): **141**:237 **(59.49%)**
Damage D:T (D/T): __**7,066**__:6,496 (1.09)
Average life duration (damage/life): 24s (353.3)
Captures: 7
Occupation time: 1m 32s
Secures: 2
Offensive kills: 4
Defensive kills: __**5**__
<:TripleKill:1322814228477906944> 4x<:DoubleKill:1322814147980689430> <:KillingSpree:1322803050347499541> <:Perfect:1322879561515532308> <:FromtheGrave:1322884008773816351> <:Rifleman:1322860656399220828> <:Stick:1322871787695898676> 3x<:Reversal:1322872797701083166> <:Boxer:1322864064082477077> <:Bodyguard:1322867380136841279>",
    },
    {
      "inline": true,
      "name": "gamertag05 (Rank: 1 | Score: 2,925)",
      "value": "Kills: 20
Deaths: 17
Assists: 10
KDA: 6.33
Headshot kills: 9
Shots H:F (acc): 128:**248** (51.61%)
Damage D:T (D/T): 6,660:6,197 (1.07)
Average life duration (damage/life): 28s (391.76)
Captures: **8**
Occupation time: **1m 49s**
Secures: 1
Offensive kills: __**7**__
Defensive kills: 1
<:Stick:1322871787695898676> <:FromtheGrave:1322884008773816351> <:Reversal:1322872797701083166> <:Rifleman:1322860656399220828> <:Boxer:1322864064082477077> <:Wingman:1322854221095108609>",
    },
    {
      "inline": false,
      "name": "
",
      "value": "
",
    },
    {
      "inline": true,
      "name": "gamertag03 (Rank: 1 | Score: 2,855)",
      "value": "Kills: 19
Deaths: __**12**__
Assists: 9
KDA: 10
Headshot kills: 12
Shots H:F (acc): 112:229 (48.9%)
Damage D:T (D/T): 5,374:5,154 (1.04)
Average life duration (damage/life): __**43s**__ __**(447.83)**__
Captures: **8**
Occupation time: 1m 41s
Secures: __**3**__
Offensive kills: 2
Defensive kills: 4
<:KillingSpree:1322803050347499541> 2x<:DoubleKill:1322814147980689430> <:Perfect:1322879561515532308> <:BackSmack:1322872596135546891> <:Rifleman:1322860656399220828> <:LastShot:1322884170606972959> <:Bodyguard:1322867380136841279> <:GuardianAngel:1322873057928286248> <:Boxer:1322864064082477077>",
    },
    {
      "inline": true,
      "name": "gamertag04 (Rank: 1 | Score: 2,410)",
      "value": "Kills: 14
Deaths: 13
Assists: __**13**__
KDA: 5.33
Headshot kills: 11
Shots H:F (acc): 125:228 (54.82%)
Damage D:T (D/T): 5,803:__**4,939**__ __**(1.17)**__
Average life duration (damage/life): 39s (446.38)
Captures: 6
Occupation time: 1m 5s
Secures: 2
Offensive kills: 1
Defensive kills: 2
<:ShotCaller:1322884687693086771> <:Perfect:1322879561515532308> <:BackSmack:1322872596135546891> <:Rifleman:1322860656399220828> <:Wingman:1322854221095108609>",
    },
    {
      "inline": false,
      "name": "
",
      "value": "
",
    },
  ],
  "title": "Strongholds: Fake name for asset 309253f8-7a75-48ff-83e1-e7fb3db2ac47",
  "url": "https://halodatahive.com/Infinite/Match/099deb74-3f60-48cf-8784-0strongholds",
}
`;
